File: 20.valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()
        match_dict = {")":"(", "}": "{", "]": "["}
        for bracket in s:
            if bracket == ")" or bracket == "}" or bracket ==  "]":
                if len(stack) == 0:
                    return False
                l = stack.pop()
                if match_dict[bracket] != l:
                    return False
            else:
                stack.append(bracket)
        

        return len(stack) == 0

        

File: 20-valid-parentheses/20-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        map<char, char> m;
        m['('] = ')';
        m['['] = ']';
        m['{'] = '}';
        for (auto x: s) {
            if (x == '(' || x == '{'|| x == '[') {
                st.push(x);
            } else {
                if (st.empty()) return false;
                char y = st.top(); st.pop();
                if (m[y] != x) return false;
            }
        }
        if (!st.empty()) return false;
        return true;
        
    }
};